File: src/services/transport/api/mcp.ts
/**
 * MCP API module.
 * Handles Model Context Protocol server lifecycle and tool invocation.
 */

import { get, post, put, del } from './client';
import type { McpServerId } from '../types/ids';
import type {
  NewMcpServer,
  UpdateMcpServer,
  McpServerInfo,
  McpToolResult,
  ResolutionStatus,
  McpTestResult,
} from '../types/mcp';

/**
 * List all configured MCP servers with their status.
 */
export async function listMcpServers(): Promise<McpServerInfo[]> {
  return get<McpServerInfo[]>('/api/mcp/servers');
}

/**
 * Add a new MCP server configuration.
 */
export async function addMcpServer(server: NewMcpServer): Promise<McpServerInfo> {
  // Convert NewMcpServer to CreateMcpServerRequest format expected by backend
  const request = {
    name: server.name,
    server_type: server.server_type,
    command: server.config.command || undefined,
    args: server.config.args || [],
    working_dir: server.config.working_dir || undefined,
    path_extra: server.config.path_extra || undefined,
    url: server.config.url || undefined,
    env: server.env.map(e => [e.key, e.value] as [string, string]),
    lifecycle: server.lifecycle,
  };
  return post<McpServerInfo>('/api/mcp/servers', request);
}

/**
 * Update an existing MCP server configuration.
 */
export async function updateMcpServer(
  id: McpServerId,
  updates: UpdateMcpServer
): Promise<McpServerInfo> {
  // Convert UpdateMcpServer to UpdateMcpServerRequest format expected by backend
  const request: Record<string, unknown> = {};
  if (updates.name !== undefined) request.name = updates.name;
  if (updates.config?.command !== undefined) request.command = updates.config.command;
  if (updates.config?.args !== undefined) request.args = updates.config.args;
  if (updates.config?.working_dir !== undefined) request.working_dir = updates.config.working_dir;
  if (updates.config?.path_extra !== undefined) request.path_extra = updates.config.path_extra;
  if (updates.config?.url !== undefined) request.url = updates.config.url;
  if (updates.env !== undefined) {
    request.env = updates.env.map(e => [e.key, e.value] as [string, string]);
  }
  if (updates.enabled !== undefined) request.enabled = updates.enabled;
  if (updates.lifecycle !== undefined) request.lifecycle = updates.lifecycle;
  
  return put<McpServerInfo>(`/api/mcp/servers/${id}`, request);
}

/**
 * Remove an MCP server configuration.
 */
export async function removeMcpServer(id: McpServerId): Promise<void> {
  await del<void>(`/api/mcp/servers/${id}`);
}

/**
 * Start an MCP server, answering with its server info — including the tools
 * the started instance advertises.
 */
export async function startMcpServer(id: McpServerId): Promise<McpServerInfo> {
  return post<McpServerInfo>(`/api/mcp/servers/${id}/start`);
}

/**
 * Stop an MCP server, answering with its server info in the stopped state.
 */
export async function stopMcpServer(id: McpServerId): Promise<McpServerInfo> {
  return post<McpServerInfo>(`/api/mcp/servers/${id}/stop`);
}

/**
 * Call an MCP tool on a specific server.
 * 
 * Note: The backend returns {success, data, error} but the HTTP client's readData()
 * function unwraps the `data` field. We handle both wrapped and unwrapped responses.
 */
export async function callMcpTool(
  serverId: McpServerId,
  toolName: string,
  args: Record<string, unknown>
): Promise<McpToolResult> {
  try {
    const result = await post<unknown>('/api/mcp/tools/call', {
      server_id: serverId,
      tool_name: toolName,
      arguments: args,
    });
    
    // Check if result is already the full McpToolResult structure
    if (typeof result === 'object' && result !== null && 'success' in result) {
      return result as McpToolResult;
    }
    
    // Result was unwrapped by readData() - it's just the data field
    // This means the call succeeded (otherwise readData would have thrown)
    return {
      success: true,
      data: result,
      // `null`, not `undefined`: the handler builds this field on both paths,
      // so a successful call sends `"error": null` and the client's own
      // success value has to look like the one the wire produces.
      error: null,
    };
  } catch (error) {
    // Network or HTTP error - convert to McpToolResult format
    const message = error instanceof Error ? error.message : String(error);
    return {
      success: false,
      // `null` for the same reason `error` is on the success path: the handler
      // builds both fields on both paths, so a real failure carries
      // `"data": null`. `undefined` serialises to an absent key — a third
      // shape neither side describes.
      data: null,
      error: message,
    };
  }
}

/**
 * Resolve MCP server executable path (for diagnostics/auto-fix).
 * Returns resolution status with success flag and detailed attempts.
 */
export async function resolveMcpServerPath(id: McpServerId): Promise<ResolutionStatus> {
  return post<ResolutionStatus>(`/api/mcp/servers/${id}/resolve`, {});
}

/**
 * Test a server's stored configuration — `gglib mcp test`.
 *
 * Starts a throwaway instance, lists its tools, stops it. Unlike starting the
 * server for real, this answers "is this config right?" without leaving a
 * process running, and it is the only way to find out short of a chat that
 * silently has no tools.
 */
export async function testMcpServer(id: McpServerId): Promise<McpTestResult> {
  return post<McpTestResult>(`/api/mcp/servers/${id}/test`, {});
}

File: crates/gglib-axum/src/handlers/mcp.rs
//! MCP handlers - Model Context Protocol server management.

use axum::Json;
use axum::extract::{Path, State};
use serde::Deserialize;

use crate::error::HttpError;
use crate::state::AppState;
use gglib_app_services::types::{
    CreateMcpServerRequest, McpServerInfo, McpTestResult, McpToolCallRequest, McpToolCallResponse,
    UpdateMcpServerRequest,
};

/// List all MCP servers.
pub(crate) async fn list(
    State(state): State<AppState>,
) -> Result<Json<Vec<McpServerInfo>>, HttpError> {
    Ok(Json(state.mcp_ops.list().await?))
}

/// Add a new MCP server.
pub(crate) async fn add(
    State(state): State<AppState>,
    Json(req): Json<CreateMcpServerRequest>,
) -> Result<Json<McpServerInfo>, HttpError> {
    Ok(Json(state.mcp_ops.add(req).await?))
}

/// Update an MCP server.
pub(crate) async fn update(
    State(state): State<AppState>,
    Path(id): Path<i64>,
    Json(req): Json<UpdateMcpServerRequest>,
) -> Result<Json<McpServerInfo>, HttpError> {
    Ok(Json(state.mcp_ops.update(id, req).await?))
}

/// Remove an MCP server.
pub(crate) async fn remove(
    State(state): State<AppState>,
    Path(id): Path<i64>,
) -> Result<(), HttpError> {
    state.mcp_ops.remove(id).await?;
    Ok(())
}

/// Start an MCP server.
pub(crate) async fn start(
    State(state): State<AppState>,
    Path(id): Path<i64>,
) -> Result<Json<McpServerInfo>, HttpError> {
    Ok(Json(state.mcp_ops.start(id).await?))
}

/// Stop an MCP server.
pub(crate) async fn stop(
    State(state): State<AppState>,
    Path(id): Path<i64>,
) -> Result<Json<McpServerInfo>, HttpError> {
    Ok(Json(state.mcp_ops.stop(id).await?))
}

/// Test a server's stored configuration end to end — `gglib mcp test`.
///
/// Starts a throwaway instance, lists its tools, stops it. A failed
/// connection returns 200 with `ok: false` and the reason: a bad command is
/// the ordinary case this exists to diagnose, and the caller wants to render
/// it beside the config rather than handle an error.
pub(crate) async fn test_connection(
    State(state): State<AppState>,
    Path(id): Path<i64>,
) -> Result<Json<McpTestResult>, HttpError> {
    Ok(Json(state.mcp_ops.test_connection(id).await?))
}

/// Request body for calling an MCP tool (includes server ID).
#[derive(Debug, Deserialize)]
pub(crate) struct CallToolRequest {
    pub server_id: i64,
    #[serde(flatten)]
    pub call: McpToolCallRequest,
}

/// Call a tool on an MCP server.
pub(crate) async fn call_tool(
    State(state): State<AppState>,
    Json(req): Json<CallToolRequest>,
) -> Result<Json<McpToolCallResponse>, HttpError> {
    Ok(Json(
        state.mcp_ops.call_tool(req.server_id, req.call).await?,
    ))
}

/// Resolve MCP server executable path (for diagnostics/auto-fix).
///
/// Returns 200 with success:false for resolution failures (not a 404/500).
pub(crate) async fn resolve_path(
    State(state): State<AppState>,
    Path(id): Path<i64>,
) -> Result<Json<gglib_core::ports::ResolutionStatus>, HttpError> {
    Ok(Json(state.mcp_ops.resolve_path(id).await?))
}
